grudge/trace_pair.py

```diff
@@ -272,61 +277,130 @@ def connected_ranks(dcoll: DiscretizationCollection):
     return get_connected_partitions(dcoll._volume_discr.mesh)
 
 
+def _flatten_to_numpy_from_container(
+        actx: ArrayContext, ary: ArrayOrContainerT) -> ArrayT:
+    """Flattens an arbitrary array container *ary* into a one-dimensional
+    numpy array suitable for communication via mpi4py.
+    """
+    if is_array_container(ary):
+        flat_data = []
+        # NOTE: The ordering of data in the flattened array will be determined
+        # by the ordering of key/value pairs returned by *serialize_container*.
+        # This is important to remember because the function
+        # *_unflatten_from_numpy_to_container* relies on this same ordering
+        # to unflatten the concatenated array.
+        for _, value in serialize_container(to_numpy(ary, actx)):
+            # NOTE: *to_numpy* can return numpy object arrays whose components
+            # are numpy arrays. As a result, we need to flatten these as well.
+            if isinstance(value, np.ndarray) and value.dtype == "O":
```

Comment:
Why are object arrays handled specially here? Can't they go through the same logic as all other containers? (If not, add a comment to explain.)

Reply:
@inducer Because `serialize_container` does not serialize a nested object array inside an array container.

Reply:
From a super quick look, this doesn't seem to be recursive. Will it handle other arbitrarily nested array containers?

Reply:
@alexfikl No, but the special casing of object arrays sorta touches on this problem at bit.

Reply:
Yeah, `serialize_container` just gives the children of one node in the container. You would need to recurse into each `value` and flatten it bottom up to avoid any special casing like that.

Reply:
I had missed this initially (good catch @alexfikl!), but I agree: this should support nested containers in their full generality, i.e. it needs to be recursive.

Reply:
This PR now requires https://github.com/inducer/arraycontext/pull/91